code/ui/blocks/src/blocks/Controls.tsx

```diff
@@ -0,0 +1,98 @@
+/* eslint-disable react/destructuring-assignment */
+import type { Args, Globals, Renderer } from '@storybook/csf';
+import type { DocsContextProps, ModuleExports, PreparedStory } from '@storybook/types';
+import type { FC } from 'react';
+import React, { useCallback, useEffect, useState, useContext } from 'react';
+import type { PropDescriptor } from '@storybook/preview-api';
+import { filterArgTypes } from '@storybook/preview-api';
+import {
+  STORY_ARGS_UPDATED,
+  UPDATE_STORY_ARGS,
+  RESET_STORY_ARGS,
+  GLOBALS_UPDATED,
+} from '@storybook/core-events';
+
+import type { SortType } from '../components';
+import { ArgsTable as PureArgsTable } from '../components';
+import { DocsContext } from './DocsContext';
+
+type ControlsParameters = {
+  include?: PropDescriptor;
+  exclude?: PropDescriptor;
+  sort?: SortType;
+};
+
+type ControlsProps = ControlsParameters & {
+  of?: Renderer['component'] | ModuleExports;
+};
+
+const useArgs = (
+  story: PreparedStory,
+  context: DocsContextProps
+): [Args, (args: Args) => void, (argNames?: string[]) => void] => {
+  const storyContext = context.getStoryContext(story);
+  const { id: storyId } = story;
+
+  const [args, setArgs] = useState(storyContext.args);
+  useEffect(() => {
+    const cb = (changed: { storyId: string; args: Args }) => {
+      if (changed.storyId === storyId) {
+        setArgs(changed.args);
+      }
+    };
+    context.channel.on(STORY_ARGS_UPDATED, cb);
+    return () => context.channel.off(STORY_ARGS_UPDATED, cb);
+  }, [storyId, context.channel]);
+  const updateArgs = useCallback(
+    (updatedArgs) => context.channel.emit(UPDATE_STORY_ARGS, { storyId, updatedArgs }),
+    [storyId, context.channel]
+  );
```

Comment:
it doesn't create any problems with closures or anything, that `updateArgs` is defined twice here?

Reply:
Twice? Do you mean once in `useArgs` and once in the `Controls` block?

Reply:
no sorry, I read both as `updateArgs` but line 47 has an extra "d". never mind. 🤦